### DISCUSSION TITLE
Linter for field order in struct initialization

### DISCUSSION THREAD

**Discussion by begelundmuller** [golangci/golangci-lint#4919] [Linter Ideas] [answered]
Hello, do you know of a linter that will error if a struct is initialized with a different field order than the struct definition?

For example:
```go
// Given this struct definition:
type Foobar struct {
  Foo int
  Bar int
}

// This should succeed:
var a = Foobar{
  Foo: 3,
  Bar: 14,
}

// But this should fail:
var b = Foobar{
  Bar: 14,
  Foo: 3,
}
```

**Comment by ldez** [comment]
Hello,

I don't think a linter exists for that as it's a pure "style" linter.

Can you explain the need behind this? Why the order is important for you?

**Reply by begelundmuller** [reply]
Hello @ldez, thank you for responding!

Yes it's definitely only a style/readability matter. Here are two reasons why it would be useful:
1. Struct fields are usually ordered in a meaningful way to improve code readability. For example, the fields in [`url.URL`](https://pkg.go.dev/net/url#URL) are roughly ordered by where they appear in the string representation of a URL. I find that using the same field order when initializing a struct similarly improves the readability of the initialization code.
2. When comparing a struct initialization to its type definition, having a consistent field order makes it easier to check if a specific field is set or not. This is particularly useful when auditing large options/config/request types (for example, [`git.CloneOptions`](https://pkg.go.dev/github.com/go-git/go-git/v5#CloneOptions) or [`s3.GetObjectInput`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/s3#GetObjectInput)).

**Comment by manuelarte** [comment, marked as answer]
Hi @begelundmuller,

I created a linter that does what you want here: https://github.com/manuelarte/structinit

In the [README.md](https://github.com/manuelarte/structfieldinitorder#%EF%B8%8F--getting-started) you should find the instructions on how to run it. Let me know how it goes if you use it.

After some talks with the maintainers of golangci-lint, I don't think this linter will be included in it.

**Reply by arkriny** [reply]
Thank you for your linter, it's really helpful!
I'm also curious why it won't be included in golangci-lint. It seems quite useful, even if it's mainly about style - there are already style-focused linters included, `wsl` for example. Additionally, this linter complements the "fill struct" code action very well, which often used to initialize all fields in the declared order. Paired with `exhaustruct`, this linter can ensures readability and consistency in every place.
Even if it remains like that, it would still be great to use it as a plugin, would that be possible? Thanks again for your work!

**Reply by manuelarte** [reply]
Hi @arkriny I updated the repository and also the README to indicate how to use it as a plugin. Let me know if you find it useful.
